src/Voron/Data/Graphs/Hnsw.cs

```diff
@@ -0,0 +1,1315 @@
+                    }
+                    distances.Slice(index, total).Fill(distance);
+                    index += total;
+                    continue;
+                }
+
+                if (_currentMatchesIndex < _postingListResults.Count)
+                {
+                    var copy = Math.Min(_postingListResults.Count - _currentMatchesIndex, matches.Length - index);
+                    _postingListResults.CopyTo(matches[index..], _currentMatchesIndex, copy);
+                    distances.Slice(index, copy).Fill(distance);
+                    index += copy;
+                    _currentMatchesIndex += copy;
+                    if(_currentMatchesIndex == _postingListResults.Count)
+                    {
+                        _currentMatchesIndex = 0;
+                        _postingListResults.Clear();
+                        _currentNode++;
+                    }
+                    continue;
+                }
+
+                var nodeIdx = _indexes[_currentNode];
+                ref var node = ref _searchState.GetNodeByIndex(nodeIdx);
+                var rawPostingListId = node.PostingListId & Constants.Graphs.VectorId.ContainerType;
+                distance = _searchState.Distance(_vector.ToSpan(), -1, nodeIdx);
+                switch (node.PostingListId & Constants.Graphs.VectorId.EnsureIsSingleMask)
+                {
+                    case Constants.Graphs.VectorId.Tombstone: // empty
+                        _currentNode++;
+                        continue;
+                    case Constants.Graphs.VectorId.Single: // single item posting list
+                        distances[index] = distance;
+                        matches[index++] = rawPostingListId;
+                        _currentNode++;
+                        continue;
+                    case Constants.Graphs.VectorId.SmallPostingList: // small posting list
+                        Debug.Assert(_postingListResults.Count is 0 && _currentMatchesIndex is 0);
+                        _searchState.ReadPostingList(rawPostingListId, ref _postingListResults, ref _pforDecoder, out _);
+                        continue;
+                    case Constants.Graphs.VectorId.PostingList: // large posting list
+                        var setStateSpan = Container.GetReadOnly(_searchState.Llt, rawPostingListId);
+                        ref readonly var setState = ref MemoryMarshal.AsRef<PostingListState>(setStateSpan);
+                        _postingList = new PostingList(_searchState.Llt, Slices.Empty, setState);
+                        _postingListIterator = _postingList.Iterate();
+                        continue;
+                    default:
+                        throw new ArgumentOutOfRangeException("Impossible scenario, we have only 4 options, but got: " + node.PostingListId);
+                }
+            }
+
+            Registration.InternalEntryIdToEntryId(matches.Slice(0, index));
+            return index;
+        }
+    }
+}
```

Comment:
It's pretty big file. Can we move nested classes / structs to separate files?

Reply:
Agree, but let's do it at the end of the work since it's easier to track the history of changes during development.